tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,175 @@
+/*************************************************************************
+ * Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "ROOT/InternalTreeUtils.hxx"
+#include "TTree.h"
+#include "TChain.h"
+#include "TFile.h"
+#include "TFriendElement.h"
+
+#include <utility> // std::pair
+#include <vector>
+#include <stdexcept> // std::runtime_error
+#include <string>
+
+namespace ROOT {
+namespace Internal {
+namespace TreeUtils {
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn std::vector<std::string> GetFileNamesFromTree(const TTree &tree)
+/// \ingroup tree
+/// \brief Get and store the file names associated with the input tree.
+/// \param[in] tree The tree from which friends information will be gathered.
+/// \throws std::runtime_error If no files could be associated with the input tree.
+std::vector<std::string> GetFileNamesFromTree(const TTree &tree)
+{
+   std::vector<std::string> filenames;
+
+   const bool isChain = tree.IsA() == TChain::Class();
+   if (isChain) {
+      const TObjArray *filelist = static_cast<const TChain &>(tree).GetListOfFiles();
+      const auto nFiles = filelist->GetEntries();
+      if (nFiles == 0)
+         throw std::runtime_error("The provided chain of files is empty");
+      filenames.reserve(nFiles);
+      for (const auto *f : *filelist)
+         filenames.emplace_back(f->GetTitle());
+   } else {
+      const TFile *f = tree.GetCurrentFile();
+      if (!f) {
+         const auto msg = "The specified TTree is not linked to any file, in-memory-only trees are not supported.";
+         throw std::runtime_error(msg);
+      }
+
+      filenames.emplace_back(f->GetName());
+   }
+
+   return filenames;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn FriendInfo GetFriendInfo(const TTree &tree)
+/// \ingroup tree
+/// \brief Get and store the names, aliases and file names of the friends of the tree.
+/// \param[in] tree The tree from which friends information will be gathered.
+///
+/// \note "friends of friends" and circular references in the lists of friends
+///       are not supported.
+FriendInfo GetFriendInfo(const TTree &tree)
```

Comment:
```suggestion
FriendInfo GetDirectFriendsInfo(const TTree &tree)
```
As this does not seem to recurse through the friends of the friends.  At the very least (i.e. possibly instead) this should be noted in the documentation string.

Reply:
it's documented right above with a `\note`?

Reply:
humm I did somehow missed .. but ... is "not supported" even the right term.   There is nothing inherently "wrong" with calling this routine on a TTree which has Friend with Friend, it will "just" return information about the direct friend.  [That then the customer of that information is *also* not looking at the friend of friends is another story].  [Similarly the information is similar (but more compact) that the information returned by `TTree::GetListOfFriends`].

Reply:
Will improve the docs in the next commit